presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorWithEstimatedExchanges.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.Map;
+
+import static com.facebook.presto.cost.PlanNodeCostEstimate.ZERO_COST;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPLICATE;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This is a wrapper class around CostCalculator that estimates ExchangeNodes cost.
+ */
+@ThreadSafe
+public class CostCalculatorWithEstimatedExchanges
+        implements CostCalculator
+{
+    private final CostCalculator costCalculator;
+    private final int numberOfNodes;
+
+    @Inject
+    public CostCalculatorWithEstimatedExchanges(CostCalculator costCalculator, InternalNodeManager nodeManager)
+    {
+        this(costCalculator, nodeManager.getAllNodes().getActiveNodes().size());
+    }
+
+    public CostCalculatorWithEstimatedExchanges(CostCalculator costCalculator, int numberOfNodes)
+    {
+        this.costCalculator = requireNonNull(costCalculator, "costCalculator is null");
+        this.numberOfNodes = numberOfNodes;
+    }
+
+    @Override
+    public PlanNodeCostEstimate calculateRootCost(Session session, Map<Symbol, Type> types, PlanNode planNode, Lookup lookup)
+    {
+        ExchangeCostEstimator exchangeCostEstimator = new ExchangeCostEstimator(
+                session,
+                types,
+                lookup,
+                numberOfNodes);
+        PlanNodeCostEstimate estimatedExchangeCost = planNode.accept(exchangeCostEstimator, null);
+
+        return costCalculator.calculateRootCost(session, types, planNode, lookup).add(estimatedExchangeCost);
```

Comment:
I don't want to duplicate logic of a `CostCalculatorUsingExchanges`. It has non trivial logic, that is easy to make a mistake in it.